driver-core/src/main/java/com/datastax/driver/core/SessionManager.java

```diff
@@ -533,34 +526,52 @@ void execute(RequestHandler.Callback callback, Statement statement) {
         new RequestHandler(this, callback, statement).sendRequest();
     }
 
-    private void prepare(String query, InetSocketAddress toExclude) throws InterruptedException {
-        for (Map.Entry<Host, HostConnectionPool> entry : pools.entrySet()) {
+    private void prepare(final String query, InetSocketAddress toExclude) {
+        List<ListenableFuture<?>> futures = new ArrayList<ListenableFuture<?>>(pools.size());
+        for (final Map.Entry<Host, HostConnectionPool> entry : pools.entrySet()) {
             if (entry.getKey().getSocketAddress().equals(toExclude))
                 continue;
-
-            // Let's not wait too long if we can't get a connection. Things
-            // will fix themselves once the user tries a query anyway.
-            Connection c = null;
-            boolean timedOut = false;
-            try {
-                c = entry.getValue().borrowConnection(200, TimeUnit.MILLISECONDS);
-                c.write(new Requests.Prepare(query)).get();
-            } catch (ConnectionException e) {
-                // Again, not being able to prepare the query right now is no big deal, so just ignore
-            } catch (BusyConnectionException e) {
-                // Same as above
-            } catch (TimeoutException e) {
-                // Same as above
-            } catch (ExecutionException e) {
-                // We shouldn't really get exception while preparing a
-                // query, so log this (but ignore otherwise as it's not a big deal)
-                logger.error(String.format("Unexpected error while preparing query (%s) on %s", query, entry.getKey()), e);
-                // If the query timed out, that already released the connection
-                timedOut = e.getCause() instanceof OperationTimedOutException;
-            } finally {
-                if (c != null && !timedOut)
-                    c.release();
-            }
+            // prepare on other nodes: we use an executor so that the query is not done in an IO thread.
+            futures.add(executor().submit(new Runnable() {
+                @Override
+                public void run() {
+                    Connection c = null;
+                    boolean timedOut = false;
+                    try {
+                        // Let's not wait too long if we can't get a connection. Things
+                        // will fix themselves once the user tries a query anyway.
+                        c = entry.getValue().borrowConnection(200, TimeUnit.MILLISECONDS);
+                        c.write(new Requests.Prepare(query)).get();
```

Comment:
`c.write` returns a future, we could use it directly instead of submitting a task to `executor`.
There's still a problem with the blocking `borrowConnection` call. But since these calls are not critical, it's acceptable to borrow with timeout = 0 here, and simply not send the query if no connection is available.